### PR TITLE
Tests crypto aesctr

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

File: tests/crypto_aesctr/main.c [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+	    "Ceci n'est pas 24 chars.",
+	    "b1f563df0a27b8b5da87ba1abc5d57b2c47d15c62bcbeccb"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "This block is exactly 32 chars!!",
+	    "a6f869c50a2bf3bfca98ba03ae0e12f8913e02c23399acd78695f3503ab1171c"}
+};
+
+/* Largest buffer must be first.  1028 is what's used in spiped. */
+static const size_t perfsizes[] = {16384, 8192, 4096, 2048, 1028, 1024, 512, 256,
+    128, 64, 32, 16};
```
Comment (gperciva) [inline tests/crypto_aesctr/main.c:95]: The first idea for benchmarking AES-CTR was "do the same thing as the crc32 test".  But are all these sizes relevant?  spiped only uses 1028-byte buffers.  scrypt uses variable size[1], but if somebody is only trying to encrypt 16 bytes then aesctr isn't going to be the bottleneck.  Tarsnap is variable as well.

[1] for buffers, it's the entire thing at once.  For files, it's up to 65536 at once.

FWIW, the drop-off on my desktop machine (of the un-optimized code) is:

```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 32896 blocks of size 16384:	2.003606 s	269.0 MB/s
... in 65792 blocks of size 8192:	2.002147 s	269.2 MB/s
... in 131584 blocks of size 4096:	2.009137 s	268.3 MB/s
... in 263168 blocks of size 2048:	2.003082 s	269.1 MB/s
... in 524288 blocks of size 1028:	2.025053 s	266.2 MB/s
... in 526336 blocks of size 1024:	2.005457 s	268.8 MB/s
... in 1052672 blocks of size 512:	2.015918 s	267.4 MB/s
... in 2105344 blocks of size 256:	2.041611 s	264.0 MB/s
... in 4210688 blocks of size 128:	2.094933 s	257.3 MB/s
... in 8421376 blocks of size 64:	2.205024 s	244.4 MB/s
... in 16842752 blocks of size 32:	2.419742 s	222.7 MB/s
... in 33685504 blocks of size 16:	2.854426 s	188.8 MB/s
Overall time and speed:	25.68 s	251.9 MB/s
```

Comment (cperciva) [reply tests/crypto_aesctr/main.c:95]: Maybe make it sizes 1--1024?  That seems like the interesting range.

File: tests/crypto_aesctr/generate_test_vectors.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    "Ceci n'est pas 24 chars.",
+    "This block is exactly 32 chars!!",
+    ]
+
+INITIAL_VECTOR = b'\0' * 16
+
+
+def generate_for_keylen(keylen):
+    """ Print plaintext and ciphertext for the given key length.  """
+    # key: 00 02 03 04...
```
Comment (cperciva) [inline tests/crypto_aesctr/generate_test_vectors.py:22]: 0,2,3,4?  Is that a typo?

Comment (gperciva) [reply tests/crypto_aesctr/generate_test_vectors.py:22]: Yes, a typo; the code does 0, 1, 2, etc.

File: tests/crypto_aesctr/main.c [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+		bufsize = perfsizes[i];
+		num_hashes = bytes_to_hash / bufsize;
+
+		/* Get beginning time. */
+		if (monoclock_get_cputime(&begin)) {
+			warnp("monoclock_get_cputime()");
+			goto err2;
+		}
+
+		/* Hash all the bytes. */
```
Comment (cperciva) [inline tests/crypto_aesctr/main.c:228]: `s/Hash/Encrypt/`

File: tests/crypto_aesctr/main.c (hunk 1 above)
Comment (gperciva) [reply tests/crypto_aesctr/main.c:95]: Including 1028 for testing exactly how it'll be used in spiped, or just end at 1024?

Comment (cperciva) [reply tests/crypto_aesctr/main.c:95]: Just end at 1024.  The speed for 1028 isn't much different and I don't really want libcperciva polluted by spiped-specific details.

Comment (gperciva) [reply tests/crypto_aesctr/main.c:95]: Hmm, are you sure you want to go all the way down to 1?  Like, `crypto_aesctr_buf(key_exp, 0, buf, buf, 1);`?  It's atrociously slow, but surely nobody would be calling it on a 1-byte buffer for any kind of time-sensitive task.

```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 526336 blocks of size 1024:	2.010606 s	268.1 MB/s
... in 1052672 blocks of size 512:	2.022461 s	266.5 MB/s
... in 2105344 blocks of size 256:	2.049952 s	262.9 MB/s
... in 4210688 blocks of size 128:	2.095254 s	257.2 MB/s
... in 8421376 blocks of size 64:	2.204425 s	244.5 MB/s
... in 16842752 blocks of size 32:	2.420215 s	222.7 MB/s
... in 33685504 blocks of size 16:	2.853046 s	188.9 MB/s
... in 67371008 blocks of size 8:	5.026268 s	107.2 MB/s
... in 134742016 blocks of size 4:	9.405395 s	57.3 MB/s
... in 269484032 blocks of size 2:	18.148134 s	29.7 MB/s
... in 538968064 blocks of size 1:	35.610145 s	15.1 MB/s
Overall time and speed:	83.85 s	70.7 MB/s
```

If you really want this, ok, but in that case I think we should omit the "Overall time and speed", because for any serious use we'd surely be using bigger buffer sizes.
(I'm also going to play with a smaller total byte size.)

Comment (cperciva) [reply tests/crypto_aesctr/main.c:95]: Yes, "overall time and speed" is a red herring so I'd drop that.

Yes, nobody should be encrypting one byte at a time; in a way, that's why I think it's useful to have the benchmark show how terrible that is. ;-)

I'd drop the total number of bytes, maybe down to 32 MB.

Comment (gperciva) [reply tests/crypto_aesctr/main.c]: Ok.
```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 32768 blocks of size 1024:	0.125179 s	268.1 MB/s
... in 65536 blocks of size 512:	0.125577 s	267.2 MB/s
... in 131072 blocks of size 256:	0.127058 s	264.1 MB/s
... in 262144 blocks of size 128:	0.130472 s	257.2 MB/s
... in 524288 blocks of size 64:	0.137340 s	244.3 MB/s
... in 1048576 blocks of size 32:	0.150582 s	222.8 MB/s
... in 2097152 blocks of size 16:	0.184997 s	181.4 MB/s
... in 4194304 blocks of size 8:	0.312804 s	107.3 MB/s
... in 8388608 blocks of size 4:	0.583361 s	57.5 MB/s
... in 16777216 blocks of size 2:	1.129428 s	29.7 MB/s
... in 33554432 blocks of size 1:	2.216826 s	15.1 MB/s
$
```

**Comment by cperciva** [comment]
Looks plausible.  Can you clean up the commit history?

**Comment by gperciva** [comment]
Updated, but added another REBASE commit.  Swapping the order of perftests means that if the "warm up" isn't sufficient to fully get the CPU going, it's the 1-byte array speed that will suffer, not the 1024-byte array speed.  I think that the larger buffer sizes are what we'll be looking at in terms of optimization.

(of course, if the warm up is consistently not sufficient on a particular system we're focusing on, then we can and should increase it)

**Comment by cperciva** [comment]
I'm not sure the order-reversal is needed -- if the test shows that 1024-byte buffers are slower than 512-byte buffers, it's pretty easy to guess that it's a warmup issue -- but I don't object either.

File: tests/crypto_aesctr/main.c [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+	printf("Computing %zu-bit AES-CTR of a large buffer two "
+	    "different ways...", keylen * 8);
+
+	/* Prepare output buffers. */
+	if ((largebuf_out1 = malloc(LARGE_BUFSIZE)) == NULL)
+		goto err2;
+	if ((largebuf_out2 = malloc(LARGE_BUFSIZE)) == NULL)
+		goto err3;
+
+	/* Compute checksum with one call. */
```
Comment (cperciva) [inline tests/crypto_aesctr/main.c:292]: Encrypt?

File: tests/crypto_aesctr/main.c [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+		} else {
+			printf(" PASSED!\n");
+		}
+
+		/* Clean up. */
+		crypto_aes_key_free(key_exp);
+	}
+
+	/* Test unaligned access. */
+	if(selftest_unaligned_access(16))
```
Comment (cperciva) [inline tests/crypto_aesctr/main.c:390]: Missing whitespace.

**Comment by gperciva** [comment]
Updated with REBASE commits.

File: tests/crypto_aesctr/main.c [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+	if ((largebuf_out2 = malloc(LARGE_BUFSIZE)) == NULL)
+		goto err3;
+
+	/* Encrypt with one call. */
+	crypto_aesctr_buf(key_exp, 0, largebuf, largebuf_out1, LARGE_BUFSIZE);
+
+	/* Ensure we have a repeatable pattern of random values. */
+	srandom(0);
+
+	/* Compute checksum with multiple calls. */
```
Comment (cperciva) [inline tests/crypto_aesctr/main.c:298]: checksum

File: tests/crypto_aesctr/main.c [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+		crypto_aesctr_stream(aesctr, &largebuf[bytes_processed],
+		    &largebuf_out2[bytes_processed], new_chunk);
+		bytes_processed += new_chunk;
+	}
+	new_chunk = LARGE_BUFSIZE - bytes_processed;
+	crypto_aesctr_stream(aesctr, &largebuf[bytes_processed],
+	    &largebuf_out2[bytes_processed], new_chunk);
+	crypto_aesctr_free(aesctr);
+
+	/* Compare checksums. */
```
Comment (cperciva) [inline tests/crypto_aesctr/main.c:313]: checksums

**Comment by gperciva** [comment]
Sorry.  Here's more fixes to the comments.

File: tests/crypto_aesctr/generate_test_vectors.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
+
+""" Generate AES-CTR test vectors with an independent implementation. """
+
+import binascii
+
+import Crypto.Util.Counter
```
Comment (gperciva) [inline tests/crypto_aesctr/generate_test_vectors.py:7]: In case you're wondering, the blank line separates `binascii`, which is part of the standard library, from the `Crypto` modules, which is a separate `pip install`.

**Comment by gperciva** [comment]
I went ahead and rebased this PR.  Potentially ready for merge.

**Comment by cperciva** [comment]
At some point it might be nice to add tests for non-zero nonce values, but I'm going to merge this as it is first.
